File: ch02/merge_sort.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <random>

#define RMIN    -0x8000
#define RMAX    0x7FFF

using std::cout;
using std::string; using std::vector;
using std::ifstream; using std::ofstream;


void makeRandomNumbers(string, int);
void lookRandomNumbers(string, int pos = 0, int n = 100);
template <class T> void sortRandomNumbers(T, string, string);
template <typename T> void mergeSort(vector<T>&, int, int);


int main()
{
    makeRandomNumbers("test", 1024 * 1024);
    sortRandomNumbers(mergeSort<int>, "test", "out-test");
    lookRandomNumbers("out-test", 512 * 1024, 512 * 1024);
    // takes about 3 seconds
}


void makeRandomNumbers(string filename, int n)
{
    std::random_device rd;
    std::mt19937_64 gen(rd());
    std::uniform_int_distribution<long> dis(RMIN, RMAX);

    ofstream output(filename,
            std::ios::out | std::ios::trunc | std::ios::binary);
    do {
        int buffer = dis(gen);
        output.write(reinterpret_cast<char*>(&buffer), sizeof(int));
    } while (--n != 0);
    output.close();
}


void lookRandomNumbers(string filename, int pos, int n)
{
    int buffer = 0;
    ifstream input(filename, std::ios::in);
    input.seekg(pos * 4);
    while (input.read(reinterpret_cast<char*>(&buffer), sizeof(int))
           && n--) {
        cout << buffer << '\n';
    }
    input.close();
}


template <class T>
void sortRandomNumbers(T sort, string in_file, string out_file)
{
    int buffer = 0;
    vector<int> v(0);

    // read random numbers
    ifstream input(in_file, std::ios::in);
    while (input.read(reinterpret_cast<char*>(&buffer), sizeof(int))) {
        v.push_back(buffer);
    }
    input.close();

    // sort random numbers
    sort(v, 0, v.size() - 1);

    // write random numbers
    ofstream output(out_file,
            std::ios::out | std::ios::trunc | std::ios::binary);
    for (auto i = v.begin(); i != v.end(); ++i) {
        buffer = v[i - v.begin()];
        output.write(reinterpret_cast<char*>(&buffer), sizeof(int));
    }
    output.close();
}


template <typename T>
void mergeSort(vector<T>& v, int left, int right)
{
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSort(v, left, mid);
        mergeSort(v, mid + 1, right);

        vector<T> Lv(v.begin() + left, v.begin() + mid + 1);
        vector<T> Rv(v.begin() + mid + 1, v.begin() + right + 1);

        for (int i = 0, j = 0; left <= right; ++left) {
            if (i == Lv.size()) {
                v[left] = Rv[j];
                ++j;
            } else if (j == Rv.size()) {
                v[left] = Lv[i];
                ++i;
            } else {
                if (Lv[i] < Rv[j]) {
                    v[left] = Lv[i];
                    ++i;
                } else {
                    v[left] = Rv[j];
                    ++j;
                }
            }
        }
    }
}

File: ch02/merge_sort.java

import java.io.*;
import java.lang.Integer;
import java.security.SecureRandom;
import java.util.ArrayList;

class MergeSort {
    public static void main(String[] args) {
        sortTest t = new sortTest();
        t.makeRandomNumbers("test", 1024 * 1024);
        t.sortRandomNumbers(t::mergeSort, "test", "out-test");
        t.lookRandomNumbers("out-test", 512 * 1024);
        // take about 33 seconds
    }

}

class sortTest {
    public static int iMin = -0x8000;
    public static int iMax = 0x7FFF;

    public int newRandom(int min, int max)
    {
        SecureRandom gen = new SecureRandom();
        return gen.nextInt(max - min + 1) + min;
    }

    public void makeRandomNumbers(String filename, int n)
    {
        try {
            DataOutputStream os =
                new DataOutputStream(new FileOutputStream(filename));
            while (n != 0) {
                os.writeInt(newRandom(iMin, iMax));
                --n;
            }
            os.close();
        } catch (FileNotFoundException e) {
            System.err.println("FileNotFoundException: " + e.getMessage());
        } catch (IOException e) {
            System.err.println("IOException: " + e.getMessage());
        }
    }

    public void lookRandomNumbers(String filename, int pos, int n)
    {
        try {
            DataInputStream is =
                new DataInputStream(new FileInputStream(filename));
            is.skipBytes(pos * 4);
            long len = new File(filename).length() / 4;
            for (; len != 0 && n != 0; --len, --n) {
                System.out.println(is.readInt());
            }
            is.close();
        } catch (FileNotFoundException e) {
            System.err.println("FileNotFoundException: " + e.getMessage());
        } catch (EOFException e) {
            System.err.println("EOFException: " + e.getMessage());
        } catch (IOException e) {
            System.err.println("IOException: " + e.getMessage());
        }
    }

    public void lookRandomNumbers(String filename, int pos)
    {
        long len = new File(filename).length() / 4;
        lookRandomNumbers(filename, pos, (int)len);
    }

    public void lookRandomNumbers(String filename)
    {
        long len = new File(filename).length() / 4;
        lookRandomNumbers(filename, 0, (int)len);
    }

    public <T> ArrayList<T> arrayCopy(ArrayList<T> v, int start, int n)
    {
        ArrayList<T> tmp = new ArrayList<>();
        for (; start != v.size() && n != 0; ++start, --n) {
            tmp.add(v.get(start));
        }
        return tmp;
    }

    public <T> void printAL(ArrayList<T> v)
    {
        for (int i = 0; i != v.size(); ++i) {
            System.out.print(" " + v.get(i));
        }
        System.out.print('\n');
    }

    public void mergeSort(ArrayList<Integer> v, int left, int right)
    {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(v, left, mid);
            mergeSort(v, mid + 1, right);

            ArrayList<Integer> La = arrayCopy(v, left, (mid - left + 1));
            ArrayList<Integer> Ra = arrayCopy(v, mid + 1, (right - mid));

            for (int i = 0, j = 0; left <= right; ++left) {
                if (i == La.size()) {
                    v.set(left, Ra.get(j));
                    ++j;
                } else if (j == Ra.size()) {
                    v.set(left, La.get(i));
                    ++i;
                } else {
                    if (La.get(i) < Ra.get(j)) {
                        v.set(left, La.get(i));
                        ++i;
                    } else {
                        v.set(left, Ra.get(j));
                        ++j;
                    }
                }
            }
        }
    }

    protected interface refSort<T>
    {
        void sort(ArrayList<T> s, int left, int rigt);
    }

    public void sortRandomNumbers(refSort<Integer> s, String input, String output)
    {
        ArrayList<Integer> v = new ArrayList<Integer>();

        // read random numbers
        try {
            DataInputStream is =
                new DataInputStream(new FileInputStream(input));
            long len = new File(input).length() / 4;
            for (; len != 0; --len) {
                v.add(is.readInt());
            }
            is.close();
        } catch (FileNotFoundException e) {
            System.err.println("FileNotFoundException: " + e.getMessage());
        } catch (EOFException e) {
            System.err.println("EOFException: " + e.getMessage());
        } catch (IOException e) {
            System.err.println("IOException: " + e.getMessage());
        }

        // sort random numbers
        s.sort(v, 0, v.size() - 1);

        // write random numbers
        try {
            DataOutputStream os =
                new DataOutputStream(new FileOutputStream(output));
            for (int i = 0; i != v.size(); ++i) {
                os.writeInt(v.get(i));
            }
            os.close();
        } catch (FileNotFoundException e) {
            System.err.println("FileNotFoundException: " + e.getMessage());
        } catch (IOException e) {
            System.err.println("IOException: " + e.getMessage());
        }
    }
}
